zerver/tests/test_subs.py

```diff
@@ -4722,27 +4722,49 @@ def test_bulk_subscribe_MIT(self) -> None:
             stream.is_in_zephyr_realm = True
             stream.save()
 
-        # Make sure Zephyr mirroring realms such as MIT do not get
-        # any tornado subscription events
         events: List[Mapping[str, Any]] = []
-        with self.tornado_redirected_to_list(events, expected_num_events=0):
-            with self.assert_database_query_count(5):
+        with self.tornado_redirected_to_list(events, expected_num_events=41):
+            with self.assert_database_query_count(52):
```

Comment:
Earlier user was not subscribed because of the wrong condition `len(principals)>0` in view function used for adding subscriptions. Now after changing the condition user is successfully subscribed to streams because of which query count increases by this much.

Reply:
41 events and 52 databases calls sounds like a performance problem.

But the reason for the Zephyr detail is that the security model for Zephyr (the protocol we're mirroring with) is that stream names have a security-through-obscurity model; anyone can subscribe if they know the name.

So we definitely intend not to send any events to other users when adjusting subscriptions in such a stream.

Reply:
Reading this, I cannot figure out what change in your code caused this to generate 41 events; I would have expected  a single event here.

Reply:
The increase in query counts and events is because of calling this function while adding subscriptions.
```
def send_stream_creation_events_for_private_streams(
    realm: Realm,
    stream_dict: Dict[int, Stream],
    altered_user_dict: Dict[int, Set[int]],
) -> None:
    for stream_id, stream_users_ids in altered_user_dict.items():
        stream = stream_dict[stream_id]

        if not stream.is_public():
            # Users newly added to invite-only streams
            # need a `create` notification.  The former, because
            # they need the stream to exist before
            # they get the "subscribe" notification, and the latter so
            # they can manage the new stream.
            # Realm admins already have all created private streams.
            realm_admin_ids = {user.id for user in realm.get_admin_users_and_bots()}
            notify_user_ids = list(stream_users_ids - realm_admin_ids)

            if notify_user_ids:
                send_stream_creation_event(stream, notify_user_ids)
```
For this test if block is executed as `stream.is_public()` returns `False` because of stream is in `zephyr_realm`. I think 
we should change the if condition to `stream.invite_only` and also move `realm_admin_ids` outside the loop.

Reply:
It appears these are `stream_created` events that get sent when the user themself joins the stream; I think that probably is correct that they get those.

So the fix here is twofold:
* We should explain that these events are from `send_stream_creation_events_for_private_streams` in the comment.
* We should assert that the events have the right `type`

Reply:
OK rewrote the test to explain the behavior clearly.